fast_llm/layers/language_model/head.py

```diff
@@ -336,30 +357,92 @@ def _logits_cross_entropy_forward_backward(
                 logits_scale_factor=self._logits_scale_factor,
-                target_format=TargetFormat.labels if self._config.distillation_model is None else TargetFormat.logits,
+                target_format=TargetFormat.labels,
             )
+            lm_loss = lm_loss * self._language_model_loss_factor
+        else:
+            lm_loss, lm_grad = None, None
+
+        if distillation_target is not None and self._distilation_loss_factor > 0.0:
+            if self._distillation_loss_implementation == DistillationLossImpl.reverse_kl:
+                distillation_loss, distillation_grad = reverse_kl_forward_backward(
+                    logits.flatten(0, -2),
+                    distillation_target,
+                    loss_mask,
+                    grad_output=grad_output * self._loss_coefficient * self._distilation_loss_factor,
+                    group=self._tensor_space.distributed.tensor_group if self._parallel_embeddings else None,
+                    logits_scale_factor=self._logits_scale_factor,
+                    teacher_softmax_temperature=self._config.teacher_softmax_temperature,
+                    target_format=(
+                        TargetFormat.labels if self._config.distillation_model is None else TargetFormat.logits
+                    ),
+                )
+            else:
+                distillation_loss, distillation_grad = cross_entropy_forward_backward(
+                    logits.flatten(0, -2),
+                    distillation_target,
+                    loss_mask,
+                    group=self._tensor_space.distributed.tensor_group if self._parallel_embeddings else None,
+                    grad_output=grad_output * self._loss_coefficient * self._distilation_loss_factor,
+                    implementation=self._cross_entropy_impl,
+                    logits_scale_factor=self._logits_scale_factor,
+                    target_format=TargetFormat.logits,
+                )
+
+            distillation_loss = distillation_loss * self._distilation_loss_factor
+        else:
+            distillation_loss, distillation_grad = None, None
 
         # TODO: de-allocate earlier.
         del logits
+
+        # TODO: Accumulate grads in-place to reduce memory and compute overhead.
+        grad = _add_tensors(dpo_grad, lm_grad, distillation_grad)
+
+        # TODO: Return individual losses?
+        loss = _add_tensors(dpo_loss, lm_loss, distillation_loss)
+        if self.training and losses is not None:
+            if dpo_loss is not None:
+                losses[LanguageModelLossNames.dpo_loss].append(dpo_loss.detach())
+            if distillation_loss is not None:
+                losses[LanguageModelLossNames.distillation_loss].append(distillation_loss.detach())
+            if self._config.distillation_model is not None and lm_loss is not None:
+                losses[LanguageModelLossNames.distil_lm_loss].append(lm_loss.detach())
```

Comment:
does this need a new label or can this just be lm loss?

Reply:
> language_model_loss is the final combined loss

ah, I see! Ok, that isn't very clear, but I guess why this is the case for historical reasons.